contrib/endpoints/src/api_manager/firebase_rules/firebase_request.cc

```diff
@@ -0,0 +1,403 @@
+                  "No external function calls present."
+                  " But received a response. Possible code bug");
+  }
+
+  funcs_with_result_.emplace_back(*func_call_iter_, body);
+  func_call_iter_++;
+  return Status::OK;
+}
+
+// Sets the next HTTP request that should be issued.
+Status FirebaseRequest::SetNextRequest() {
+  if (IsDone()) {
+    next_request_ = nullptr;
+    return current_status_;
+  }
+
+  Status status = Status::OK;
+
+  // While there are more functions that should be processed, check if the HTTP
+  // response for the function is already buffered. Set the next HTTP request if
+  // we find a new function and break.
+  while (!AllFunctionCallsProcessed()) {
+    if (Find(*func_call_iter_) == funcs_with_result_.end()) {
+      auto call = *func_call_iter_;
+      external_http_request_.url = call.args(0).string_value();
+      external_http_request_.method = call.args(1).string_value();
+      std::string body;
+      status =
+          utils::ProtoToJson(call.args(2), &body, utils::JsonOptions::DEFAULT);
+      if (status.ok()) {
+        external_http_request_.body = body;
+        next_request_ = &external_http_request_;
+      }
+      break;
+    }
+
+    func_call_iter_++;
+  }
+
+  // If All functions are processed, then issue a TestRulesetRequest.
+  if (AllFunctionCallsProcessed()) {
+    next_request_ = &firebase_http_request_;
+    return UpdateRulesetRequestBody(response_.test_results(0).function_calls());
+  }
+
+  return status;
+}
+
+Status FirebaseRequest::CheckFuncCallArgs(const FunctionCall &func) {
+  if (func.function().empty()) {
+    return Status(Code::INVALID_ARGUMENT, "No function name provided");
+  }
+
+  // We only support functions that call with three argument: HTTP URL, HTTP
+  // method and body. The body can be empty
+  if (func.args_size() < 2 || func.args_size() > 3) {
+    std::ostringstream os;
+    os << func.function() << " Require 2 or 3 arguments. But has "
+       << func.args_size();
+    return Status(Code::INVALID_ARGUMENT, os.str());
+  }
+
+  if (func.args(0).kind_case() != google::protobuf::Value::kStringValue ||
+      func.args(1).kind_case() != google::protobuf::Value::kStringValue) {
+    return Status(
+        Code::INVALID_ARGUMENT,
+        std::string(func.function() + " Arguments 1 and 2 should be strings"));
+  }
+
+  if (!utils::IsHttpRequest(func.args(0).string_value())) {
+    return Status(
+        Code::INVALID_ARGUMENT,
+        func.function() + " The first argument should be a HTTP request");
+  }
+
+  if (std::string(func.args(1).string_value()).empty()) {
+    return Status(
+        Code::INVALID_ARGUMENT,
+        func.function() + " argument 2 [HTTP METHOD] cannot be emtpy");
+  }
+
+  return Status::OK;
+}
+
+bool FirebaseRequest::AllFunctionCallsProcessed() {
+  return func_call_iter_ == response_.test_results(0).function_calls().end();
+}
+
+Status FirebaseRequest::AddFunctionMock(proto::TestRulesetRequest *request,
```

Comment:
No This is not for testing. These are private methods in the class. The FunctionMock is the name of a message in the TestRulesetRequest protobuf.